pkg/stream/xprotocol/connpool_multiplex.go

```diff
@@ -36,47 +36,72 @@ import (
 type poolMultiplex struct {
 	*connpool
 
-	clientMux     sync.Mutex
-	activeClients sync.Map
+	clientMux              sync.Mutex
+	activeClients          []sync.Map
+	currentCheckAndInitIdx int64
+
+	shutdown bool // pool is already shutdown
 }
 
 // NewPoolMultiplex generates a multiplex conn pool
 func NewPoolMultiplex(p *connpool) types.ConnectionPool {
+	maxConns := p.Host().ClusterInfo().ResourceManager().Connections().Max()
+	if maxConns == 0 {
+		// default conn num should be 1
+		maxConns = 1
+	}
+
 	return &poolMultiplex{
 		connpool:      p,
-		activeClients: sync.Map{},
+		activeClients: make([]sync.Map, maxConns),
 	}
 }
 
-func (p *poolMultiplex) init(client *activeClientMultiplex, sub types.ProtocolName) {
+func (p *poolMultiplex) init(client *activeClientMultiplex, sub types.ProtocolName, index int) {
 	utils.GoWithRecover(func() {
 		if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
 			log.DefaultLogger.Debugf("[stream] [sofarpc] [connpool] init host %s", p.Host().AddressString())
 		}
 
 		p.clientMux.Lock()
 		defer p.clientMux.Unlock()
+
+		// if the pool is already shut down, do nothing directly return
+		if p.shutdown {
+			return
+		}
+
 		client, _ := p.newActiveClient(context.Background(), sub)
 		if client != nil {
 			client.state = Connected
-			p.activeClients.Store(sub, client)
+			client.indexInPool = index
+			p.activeClients[index].Store(sub, client)
 		} else {
-			p.activeClients.Delete(sub)
+			p.activeClients[index].Delete(sub)
 		}
+
 	}, nil)
 }
 
 // CheckAndInit init the connection pool
 func (p *poolMultiplex) CheckAndInit(ctx context.Context) bool {
-	var client *activeClientMultiplex
+	var clientIdx int64 = 0 // most use cases, there will only be 1 connection
+	if len(p.activeClients) > 1 {
+		if clientIdx = getClientIDFromDownStreamCtx(ctx); clientIdx == invalidClientID {
+			clientIdx = atomic.AddInt64(&p.currentCheckAndInitIdx, 1) % int64(len(p.activeClients))
```

Comment:
maybe better use `AddUint64`  ? because the value type of `Connections().Max()`  is `uint64`.

Reply:
invalidClientID == -1 is used to check whether the client id is set in context